### PR TITLE
Support shutdown property of spring boot

### LINKED ISSUE #2784: Support `server.shutdown` property of Spring Boot
I heard that Spring Boot has released [`2.3.0.RELEASE`](https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-2.3-Release-Notes#graceful-shutdown). And the following properties have been added to the version.

- `server.shutdown`
- `spring.lifecycle.timeout-per-shutdown-phase`

Armeria also seems to need to support this property. What do you think?

See https://docs.spring.io/spring-boot/docs/2.3.0.RELEASE/reference/html/spring-boot-features.html#boot-features-graceful-shutdown

### PR DESCRIPTION
**Pull request by heowc**
Motivation:

It is an operation to apply similarly by utilizing shutdown-related properties provided by Spring Boot.

Modification:

- When this option is enabled, `ArmeriaServerConfigurator`, which sets `ServerBuilder.gracefulShutdownTimeout (..., ...)`, is registered as a Spring Bean.
- The default timeout is 30 seconds as provided by Spring Boot.
- This uses `@Value` for 1.x compatibility, It would be nice to change this later.

Result:
- Fixed #2784

### CODE REVIEW

File: spring/boot-autoconfigure/src/test/java/com/linecorp/armeria/spring/ArmeriaGracefulShutdownConfigurationTest.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    @Inject
+    @Nullable
+    private Server server;
+
+    @Value("${spring.lifecycle.timeout-per-shutdown-phase}")
+    private Duration duration;
+
+    @Test
+    public void testGracefulShutdown() throws Exception {
+        final long startTime = System.currentTimeMillis();
```
Comment (ikhoon) [inline spring/boot-autoconfigure/src/test/java/com/linecorp/armeria/spring/ArmeriaGracefulShutdownConfigurationTest.java:59]: nit: Seems better to use `System.nanoTime()`?

File: spring/boot-autoconfigure/src/main/java/com/linecorp/armeria/spring/ArmeriaAutoConfiguration.java [hunk 2]
```diff
@@ -160,4 +165,19 @@ public Server armeriaServer(
...
+     */
+    @Bean
+    @ConditionalOnProperty("server.shutdown")
+    public ArmeriaServerConfigurator gracefulShutdownServerConfigurator(
+            @Value("${server.shutdown}") String shutdown,
+            @Value("${spring.lifecycle.timeout-per-shutdown-phase:30s}") Duration duration) {
+        return sb -> {
+            if (GRACEFUL_SHUTDOWN.equalsIgnoreCase(shutdown)) {
+                sb.gracefulShutdownTimeout(duration, duration);
+            }
```
Comment (ikhoon) [inline spring/boot-autoconfigure/src/main/java/com/linecorp/armeria/spring/ArmeriaAutoConfiguration.java:180]: Could move the `if` condition to out side of lambda?

**Comment by heowc** [comment]
Thanks for the quick review. 😄

File: spring/boot-autoconfigure/src/test/java/com/linecorp/armeria/spring/ArmeriaGracefulShutdownConfigurationTest.java (hunk 1 above)
Comment (heowc) [reply spring/boot-autoconfigure/src/test/java/com/linecorp/armeria/spring/ArmeriaGracefulShutdownConfigurationTest.java]: Fixed fbf3274

File: spring/boot-autoconfigure/src/main/java/com/linecorp/armeria/spring/ArmeriaAutoConfiguration.java (hunk 2 above)
Comment (heowc) [reply spring/boot-autoconfigure/src/main/java/com/linecorp/armeria/spring/ArmeriaAutoConfiguration.java]: Fixed fbf3274

File: spring/boot-autoconfigure/src/test/java/com/linecorp/armeria/spring/ArmeriaGracefulShutdownConfigurationTest.java [hunk 3]
```diff
@@ -56,8 +56,8 @@
 
     @Test
     public void testGracefulShutdown() throws Exception {
-        final long startTime = System.currentTimeMillis();
+        final long startTime = System.nanoTime();
         server.stop().join();
-        assertThat(System.currentTimeMillis() - startTime).isGreaterThanOrEqualTo(duration.toMillis());
+        assertThat(System.nanoTime() - startTime).isGreaterThanOrEqualTo(duration.toMillis());
```
Comment (trustin) [inline spring/boot-autoconfigure/src/test/java/com/linecorp/armeria/spring/ArmeriaGracefulShutdownConfigurationTest.java:61]: `duration.toNanos()` ?

Comment (heowc) [reply spring/boot-autoconfigure/src/test/java/com/linecorp/armeria/spring/ArmeriaGracefulShutdownConfigurationTest.java:61]: Oops... I forgot.

**Comment by heowc** [comment]
> Thanks! @heowc, I'm always thankful for your next awesome Spring Boot configurations for Armeria. 😉

I will always try to make Armeira a fantastic combination with Spring (Boot). 😉

**Pull request merged**